### PR TITLE
AI Assistant: Check if the block is in the block editor context before enabling Accept title action

### LINKED ISSUE #31192: AI Assistant: Fix "Accept title" action on P2, site editor and widgets
ref pe4Cmx-1km-p2#comment-651

On P2, when using "Accept title", the block is removed but the title is gone. That happens because the title of a P2 post is defined in its content.

It is necessary to check and fix or disable for:

- P2 post
- P2 comment
- Site editor
- Widgets

### PR DESCRIPTION
**Pull request by dhasilva**
Fixes #31192

## Proposed changes:

* Adds an initial state object with the current screen to determine if the block is inserted in the block editor
* Only shows "Accept title" when the block is in the block editor

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

Block editor:
* Go to the block editor
* Add some content (tip: use the AI Assistant block to generate a haiku)
* Add an AI Assistant block
* Click on Write with AI... > Generate a post title
* Check that the `Accept title` button is displayed when the title is generated and clicking it applies the title as normal

Site editor
* Go to the site editor (wp-admin > Appearance > Editor)
* Repeat the process of the block editor
* Check that the `Accept title` button is replaced by the `Accept` button and clicking it adds the title inline like other content

P2
* Set up a P2 test site: https://github.com/Automattic/p2#set-up-a-sandboxed-test-site
* Apply this PR to your sandbox and enable the redirect on your hosts file
* In your sandboxed P2 test site, create a post
* Repeat the previous process of creating a title
* Check that the `Accept title` button is replaced by the `Accept` button and clicking it adds the title inline like other content

### CODE REVIEW

File: projects/plugins/jetpack/extensions/blocks/ai-assistant/ai-assistant.php [hunk 1]
```diff
@@ -28,6 +28,54 @@ function register_block() {
...
+ * @return string
+ */
+function render_initial_state() {
+	$initial_state = get_initial_state();
+
+	if ( empty( $initial_state['currentScreen'] ) ) {
+		return null;
+	}
+
+	return 'var jetpackAIAssistantInitialState=JSON.parse(decodeURIComponent("' . rawurlencode( wp_json_encode( $initial_state ) ) . '"));';
```
Comment (retrofox) [inline projects/plugins/jetpack/extensions/blocks/ai-assistant/ai-assistant.php]: Can we store these data in the `Jetpack_Editor_Initial_State` global object? For instance, [here we expose AI Assistant feature data](https://github.com/Automattic/jetpack/blob/c6e4509b63b6f074abcfd713a35f2d2d13993615/projects/plugins/jetpack/class.jetpack-gutenberg.php#L713).

https://github.com/Automattic/jetpack/blob/878ece2673a0d54f749a2e0436aaa8597380829a/projects/plugins/jetpack/class.jetpack-gutenberg.php#L713

image

File: projects/plugins/jetpack/extensions/blocks/ai-assistant/ai-control.js [hunk 2]
```diff
@@ -28,6 +28,8 @@ import origamiPlane from './icons/origami-plane';
 import { isUserConnected } from './lib/connection';
 import UpgradePrompt from './upgrade-prompt';
 
+const isInBlockEditor = window?.Jetpack_Editor_Initial_State?.screenBase === 'post';
```
Comment (retrofox) [inline projects/plugins/jetpack/extensions/blocks/ai-assistant/ai-control.js:31]: I don't know the API, so wondering how reliable this condition is. I was taking a look at the [is_block_editor()](https://developer.wordpress.org/reference/classes/wp_screen/is_block_editor/) method. Just FYI.

**Review by retrofox — APPROVED** [review]
LGTM. Nice one. TIL about WP_Screen 🤦

File: projects/plugins/jetpack/extensions/blocks/ai-assistant/ai-control.js (hunk 2 above)
Comment (dhasilva) [reply projects/plugins/jetpack/extensions/blocks/ai-assistant/ai-control.js:31]: Is block editor returns `true` in the site editor.

**Pull request merged**